### ISSUE TITLE
[Documentation Request] Inference example with openai/whisper models

### ISSUE THREAD

**Issue by DK013** [microsoft/onnxruntime#15235]
some examples of how to run inference with openai/whisper models will be really appreciated. I'm new to this ML stuff and having a hard time getting it done myself. I converted the pytorch models to onnx using optimum which resulted in two seperate models for encoding and decoding. Studying the original whisper repo it's hard to understand how to get the same logic or workflow replicated sing onnxruntime with 2 models and aslo some issues suggest that whisper onnx models with directml has reduced performence without kv cache and with dynamic axis. Some info about that will be appreciated also since using the models with directx is what I'm aiming for. Any example in either C# or js with onnexruntime-web will be really helpful.

Thank you

**Comment by guschmue** [comment]
If you can use python the code below should work.
For everything else you'd need to implement beam search and pre-processing. There is a good example here:
https://xenova.github.io/transformers.js/

In 1-2 weeks we'll have some model that combines everything into a single model.

```
import argparse
import numpy as np
from transformers import WhisperProcessor, WhisperForConditionalGeneration
from optimum.onnxruntime import ORTModelForSpeechSeq2Seq
from transformers import AutoProcessor, pipeline
from datasets import load_dataset


def run_whisper(model_name, use_onnx=False):
    # load model and processor
    if use_onnx:
        model_path = f"optimum/whisper-{model_name}.en"
        # processor = AutoProcessor.from_pretrained(model_path)
        processor = WhisperProcessor.from_pretrained(model_path)
        model = ORTModelForSpeechSeq2Seq.from_pretrained(model_path)
        model.config.forced_decoder_ids = None
    else:
        model_path = f"openai/whisper-{model_name}.en"
        processor = WhisperProcessor.from_pretrained(model_path)
        model = WhisperForConditionalGeneration.from_pretrained(model_path)
        model.config.forced_decoder_ids = None

    # load dummy dataset and read audio files
    ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")

    for i in range(0, 5):
        sample = ds[i]["audio"]
        input_features = processor(sample["array"], sampling_rate=sample["sampling_rate"], return_tensors="pt").input_features

        # generate token ids
        predicted_ids = model.generate(input_features)
        print(ds[i]["text"])

        transcription = processor.batch_decode(predicted_ids, skip_special_tokens=True)
        print(transcription)
        print("--")


def get_args():
    """Parse commandline."""
    parser = argparse.ArgumentParser()
    parser.add_argument("--model", default="tiny", help="path to model")
    parser.add_argument("--onnx", action="store_true", help="use onnxruntime")
    args = parser.parse_args()
    return args


def main():
    args = get_args()
    run_whisper(args.model, args.onnx)

if __name__ == "__main__":
    main()
```

**Comment by DK013** [comment]
> If you can use python the code below should work. For everything else you'd need to implement beam search and pre-processing. There is a good example here: https://xenova.github.io/transformers.js/
> 
> In 1-2 weeks we'll have some model that combines everything into a single model.
> 
> ```
> import argparse
> import numpy as np
> from transformers import WhisperProcessor, WhisperForConditionalGeneration
> from optimum.onnxruntime import ORTModelForSpeechSeq2Seq
> from transformers import AutoProcessor, pipeline
> from datasets import load_dataset
> 
> 
> def run_whisper(model_name, use_onnx=False):
>     # load model and processor
>     if use_onnx:
>         model_path = f"optimum/whisper-{model_name}.en"
>         # processor = AutoProcessor.from_pretrained(model_path)
>         processor = WhisperProcessor.from_pretrained(model_path)
>         model = ORTModelForSpeechSeq2Seq.from_pretrained(model_path)
>         model.config.forced_decoder_ids = None
>     else:
>         model_path = f"openai/whisper-{model_name}.en"
>         processor = WhisperProcessor.from_pretrained(model_path)
>         model = WhisperForConditionalGeneration.from_pretrained(model_path)
>         model.config.forced_decoder_ids = None
> 
>     # load dummy dataset and read audio files
>     ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
> 
>     for i in range(0, 5):
>         sample = ds[i]["audio"]
>         input_features = processor(sample["array"], sampling_rate=sample["sampling_rate"], return_tensors="pt").input_features
> 
>         # generate token ids
>         predicted_ids = model.generate(input_features)
>         print(ds[i]["text"])
> 
>         transcription = processor.batch_decode(predicted_ids, skip_special_tokens=True)
>         print(transcription)
>         print("--")
> 
> 
> def get_args():
>     """Parse commandline."""
>     parser = argparse.ArgumentParser()
>     parser.add_argument("--model", default="tiny", help="path to model")
>     parser.add_argument("--onnx", action="store_true", help="use onnxruntime")
>     args = parser.parse_args()
>     return args
> 
> 
> def main():
>     args = get_args()
>     run_whisper(args.model, args.onnx)
> 
> if __name__ == "__main__":
>     main()
> ```

I;'m trying to avoid python since in my usecase I need the whisper functionality in electron and I want to keep things on the client side. so making install with all the python runtime and torch etc is just too much to ask. I tried the C++ implementation using ggml format models from [this repository](https://github.com/ggerganov/whisper.cpp). But it only works on CPU. There's [another repository](https://github.com/Const-me/Whisper) which ports that C++ lib to work with DirectX11. But his port is made of old code and doesn't do a very good job on the realtime transcription.

Right now to me it feels the best solution will be to use onnxruntime to work in a C# console app so I can make use of DirectML to support all GPUs and use that cli discretely from my electron application.

**Comment by natke** [comment]
If you can wait a couple of weeks @DK013, you can use the single model solution that Guenther mentions above. I think this will be the easiest way for you to run in your environment.

**Comment by guschmue** [comment]
Since @DK013 wants to run this under directml - there is most likely some work on directml end to make the e2e model work.

**Comment by DK013** [comment]
@guschmue @natke I guess I have to wait. there's no other way. Thanks guys for reaching out. I'll be waiting for the news.

**Comment by DK013** [comment]
@guschmue @natke since crome 113 is gonna support webGPU without any lags needed and I heard onnxruntime is gonna release support for WebGPU as well, this gives me hope of using transformerjs directly in the electron app without needing to building a seperate C# console app. Maybe with this I can tweak transformerjs to support bigger models for whisper now. When can we expect this update? since electron already support WebGPU using chromium flags I would like to go ahead and start coding and testing as soon as possible.

**Comment by guschmue** [comment]
The PR for webgpu is here https://github.com/microsoft/onnxruntime/pull/14579.
We will try to support whisper on webgpu but we are not sure yet if all ops are fully optimized in 1.15.
And we close to have an end to end model that includes pre/post processing for whisper, basically raw pcm in, text out.

For tokenizers: ort-web ships with most tokenizers by including the tokenizers from onnxruntime-extensions. But the way this works right now is that the tokenizers are called in graph what makes the api mapping from huggingface hard. We might fix that in the near future by allowing to call the tokenizers out of graph via api call.
And we need to pull in a newer version for onnxruntime-extensions to get the tokenizer/decoder for whisper which would be in the next release.

**Comment by guschmue** [comment]
forgot to update this issue: whisper for ort-web with some instructions are now here:
https://github.com/microsoft/onnxruntime-inference-examples/tree/main/js/ort-whisper

**Comment by nums11** [comment]
Closing as resolved.

**Comment by csukuangfj** [comment]
I would suggest trying http://github.com/k2-fsa/sherpa-onnx
It supports running whisper models with onnxruntime in C++ as well as in Python.

You can find the documentation at
https://k2-fsa.github.io/sherpa/onnx/pretrained_models/whisper/index.html

![image](https://github.com/microsoft/onnxruntime/assets/5284924/9fbfec15-6a39-407b-b3f2-46af40d7f3e7)


You can also try it from within your browser without installing anything by visiting
https://huggingface.co/spaces/k2-fsa/automatic-speech-recognition

![image](https://github.com/microsoft/onnxruntime/assets/5284924/6afd7858-ed92-408f-b196-abd62cb40539)

---

There is also a colab notebook for it:
https://github.com/k2-fsa/colab/blob/master/sherpa-onnx/sherpa_onnx_whisper_models.ipynb
